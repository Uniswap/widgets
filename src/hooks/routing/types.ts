export enum RouterPreference {
  API = 'api',
  CLIENT = 'client',
}

export enum QuoteType {
  TRADE = 'trade',
  PRICE = 'price',
  SKIP = 'skip',
}

export type QuoteConfig =
  | { type: QuoteType.TRADE; preference: RouterPreference }
  | { type: QuoteType.PRICE; preference: RouterPreference }
  | { type: QuoteType.SKIP }

export enum PoolType {
  V2Pool = 'v2-pool',
  V3Pool = 'v3-pool',
}

// swap router API special cases these strings to represent native currencies
// all chains have "ETH" as native currency symbol except for polygon
export enum SwapRouterNativeAssets {
  MATIC = 'MATIC',
  ETH = 'ETH',
}
