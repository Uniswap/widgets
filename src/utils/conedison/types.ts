export type Address = string
export type AddressTo<T> = Record<Address, T>
export type Mutable<T> = { -readonly [P in keyof T]: T[P] }
export type Nullable<T> = T | null
export type Nullish<T> = Nullable<T> | undefined
export type Primitive = number | string | boolean | bigint | symbol | null | undefined
