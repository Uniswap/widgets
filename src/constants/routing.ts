// a list of tokens by chain
import { NativeCurrency, Token } from '@uniswap/sdk-core'

import { SupportedChainId } from './chains'
import {
  AMPL,
  CEUR_CELO,
  CMC02_CELO,
  CUSD_CELO,
  DAI,
  DAI_ARBITRUM_ONE,
  DAI_OPTIMISM,
  DAI_POLYGON,
  ETH2X_FLI,
  FEI,
  FRAX,
  FXS,
  nativeOnChain,
  PORTAL_ETH_CELO,
  PORTAL_USDC_CELO,
  renBTC,
  rETH2,
  sETH2,
  SWISE,
  TRIBE,
  USDC_BNB_CHAIN,
  USDC_MAINNET,
  USDC_POLYGON,
  USDT,
  USDT_ARBITRUM_ONE,
  USDT_BNB_CHAIN,
  USDT_OPTIMISM,
  USDT_POLYGON,
  WBTC,
  WBTC_ARBITRUM_ONE,
  WBTC_OPTIMISM,
  WETH_POLYGON,
  WRAPPED_NATIVE_CURRENCY,
} from './tokens'

type ChainTokenList = {
  readonly [chainId: number]: Array<Token | NativeCurrency>
}

const WRAPPED_NATIVE_CURRENCIES_ONLY: ChainTokenList = Object.fromEntries(
  Object.entries(WRAPPED_NATIVE_CURRENCY)
    .map(([key, value]) => [key, [value]])
    .filter(Boolean)
)

// used to construct intermediary pairs for trading
export const BASES_TO_CHECK_TRADES_AGAINST: ChainTokenList = {
  ...WRAPPED_NATIVE_CURRENCIES_ONLY,
  [SupportedChainId.MAINNET]: [
    nativeOnChain(SupportedChainId.MAINNET),
    DAI,
    USDC_MAINNET,
    USDT,
    WBTC,
    ...WRAPPED_NATIVE_CURRENCIES_ONLY[SupportedChainId.MAINNET],
  ],
  [SupportedChainId.OPTIMISM]: [
    ...WRAPPED_NATIVE_CURRENCIES_ONLY[SupportedChainId.OPTIMISM],
    DAI_OPTIMISM,
    USDT_OPTIMISM,
    WBTC_OPTIMISM,
  ],
  [SupportedChainId.ARBITRUM_ONE]: [
    ...WRAPPED_NATIVE_CURRENCIES_ONLY[SupportedChainId.ARBITRUM_ONE],
    DAI_ARBITRUM_ONE,
    USDT_ARBITRUM_ONE,
    WBTC_ARBITRUM_ONE,
  ],
  [SupportedChainId.POLYGON]: [
    ...WRAPPED_NATIVE_CURRENCIES_ONLY[SupportedChainId.POLYGON],
    DAI_POLYGON,
    USDC_POLYGON,
    USDT_POLYGON,
    WETH_POLYGON,
  ],
  [SupportedChainId.CELO]: [
    ...WRAPPED_NATIVE_CURRENCIES_ONLY[SupportedChainId.CELO],
    CUSD_CELO,
    CEUR_CELO,
    CMC02_CELO,
    PORTAL_USDC_CELO,
    PORTAL_ETH_CELO,
  ],
  [SupportedChainId.BNB]: [
    nativeOnChain(SupportedChainId.BNB),
    USDC_BNB_CHAIN,
    USDT_BNB_CHAIN,
    ...WRAPPED_NATIVE_CURRENCIES_ONLY[SupportedChainId.BNB],
  ],
}
export const ADDITIONAL_BASES: { [chainId: number]: { [tokenAddress: string]: Token[] } } = {
  [SupportedChainId.MAINNET]: {
    '0xF16E4d813f4DcfDe4c5b44f305c908742De84eF0': [ETH2X_FLI],
    [rETH2.address]: [sETH2],
    [SWISE.address]: [sETH2],
    [FEI.address]: [TRIBE],
    [TRIBE.address]: [FEI],
    [FRAX.address]: [FXS],
    [FXS.address]: [FRAX],
    [WBTC.address]: [renBTC],
    [renBTC.address]: [WBTC],
  },
}
/**
 * Some tokens can only be swapped via certain pairs, so we override the list of bases that are considered for these
 * tokens.
 */
export const CUSTOM_BASES: { [chainId: number]: { [tokenAddress: string]: Token[] } } = {
  [SupportedChainId.MAINNET]: {
    [AMPL.address]: [DAI, WRAPPED_NATIVE_CURRENCY[SupportedChainId.MAINNET] as Token],
  },
}
