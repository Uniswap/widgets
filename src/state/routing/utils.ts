import { ONE } from '@uniswap/router-sdk'
import { Currency, CurrencyAmount, Fraction, Percent } from '@uniswap/sdk-core'

export function calcMinimumAmountOut(
  slippageTolerance: Percent,
  amountOut: CurrencyAmount<Currency>
): CurrencyAmount<Currency> {
  const slippageAdjustedAmountOut = new Fraction(ONE)
    .add(slippageTolerance)
    .invert()
    .multiply(amountOut.quotient).quotient
  return CurrencyAmount.fromRawAmount(amountOut.currency, slippageAdjustedAmountOut)
}
