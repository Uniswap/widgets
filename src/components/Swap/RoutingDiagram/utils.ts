import { Protocol } from '@uniswap/router-sdk'
import { Currency, Percent } from '@uniswap/sdk-core'
import { FeeAmount } from '@uniswap/v3-sdk'
import { InterfaceTrade } from 'state/routing/types'
import { getFeeAmount } from 'utils/prices'
import { isExactInput } from 'utils/tradeType'

export interface RoutingDiagramEntry {
  percent: Percent
  path: [Currency, Currency, FeeAmount][]
  protocol: Protocol
}

/**
 * Loops through all routes on a trade and returns an array of diagram entries.
 */
export function getTokenPath(trade: InterfaceTrade): RoutingDiagramEntry[] {
  return trade.swaps.map(({ route: { path: tokenPath, pools, protocol }, inputAmount, outputAmount }) => {
    const portion = isExactInput(trade.tradeType)
      ? inputAmount.divide(trade.inputAmount)
      : outputAmount.divide(trade.outputAmount)
    const percent = new Percent(portion.numerator, portion.denominator)
    const path: RoutingDiagramEntry['path'] = []
    for (let i = 0; i < pools.length; i++) {
      const nextPool = pools[i]
      const tokenIn = tokenPath[i]
      const tokenOut = tokenPath[i + 1]
      const entry: RoutingDiagramEntry['path'][0] = [tokenIn, tokenOut, getFeeAmount(nextPool)]
      path.push(entry)
    }
    return {
      percent,
      path,
      protocol,
    }
  })
}
