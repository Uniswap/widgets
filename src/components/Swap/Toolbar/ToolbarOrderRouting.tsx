import { Trans } from '@lingui/macro'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'
import { BottomSheetModal } from 'components/BottomSheetModal'
import Column from 'components/Column'
import Popover from 'components/Popover'
import Row from 'components/Row'
import { useTooltip } from 'components/Tooltip'
import { useIsMobileWidth } from 'hooks/useIsMobileWidth'
import { useState } from 'react'
import { InterfaceTrade } from 'state/routing/types'
import styled from 'styled-components/macro'
import { ThemedText } from 'theme'
import { Body2LineHeightRem } from 'theme/type'

import RoutingDiagram, { AutoRouterHeader } from '../RoutingDiagram'

const CONTAINER_VERTICAL_PADDING_REM = 1
export const ORDER_ROUTING_HEIGHT_REM = CONTAINER_VERTICAL_PADDING_REM * 2 + Body2LineHeightRem /* Body2 line height */

const OrderRoutingRow = styled(Row)`
  height: ${ORDER_ROUTING_HEIGHT_REM}rem;
  margin: 0 1rem;
  padding: ${CONTAINER_VERTICAL_PADDING_REM}rem 0;
`

interface ToolbarOrderRoutingProps {
  trade?: InterfaceTrade
  gasUseEstimateUSD?: CurrencyAmount<Token> | null
}

export default function ToolbarOrderRouting({ trade, gasUseEstimateUSD }: ToolbarOrderRoutingProps) {
  const isMobile = useIsMobileWidth()
  const [open, setOpen] = useState(false)
  const [tooltip, setTooltip] = useState<HTMLDivElement | null>(null)
  const showTooltip = useTooltip(tooltip)
  return (
    <OrderRoutingRow flex>
      <ThemedText.Body2 color="secondary">
        <Trans>Order routing</Trans>
      </ThemedText.Body2>
      {trade &&
        (isMobile ? (
          <Row>
            <AutoRouterHeader ref={setTooltip} onClick={() => setOpen(true)} />
            <BottomSheetModal title="Route details" onClose={() => setOpen(false)} open={open}>
              <Column padded>
                <RoutingDiagram trade={trade} hideHeader />
              </Column>
            </BottomSheetModal>
          </Row>
        ) : (
          <Popover
            content={trade ? <RoutingDiagram gasUseEstimateUSD={gasUseEstimateUSD} trade={trade} /> : null}
            show={Boolean(trade) && showTooltip}
            placement="auto"
          >
            <AutoRouterHeader ref={setTooltip} />
          </Popover>
        ))}
    </OrderRoutingRow>
  )
}
