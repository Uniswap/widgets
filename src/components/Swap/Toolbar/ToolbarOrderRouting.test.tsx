import { CurrencyAmount, TradeType } from '@uniswap/sdk-core'
import { InterfaceTrade } from 'state/routing/types'
import { renderComponent } from 'test'
import { buildMultiV3Route, buildSingleV3Route, DAI, USDC } from 'test/utils'

import ToolbarOrderRouting from './ToolbarOrderRouting'

describe('ToolbarOrderRouting', () => {
  it('renders correctly, loading', () => {
    const component = renderComponent(<ToolbarOrderRouting />)
    expect(component.getByText('Order routing')).toBeTruthy()
  })

  it('renders correctly, 1 hop', () => {
    const inputAmount = CurrencyAmount.fromRawAmount(USDC, 1)
    const outputAmount = CurrencyAmount.fromRawAmount(DAI, 1)
    const testRoute = buildSingleV3Route(inputAmount, outputAmount)
    const component = renderComponent(
      <ToolbarOrderRouting
        gasUseEstimateUSD={CurrencyAmount.fromRawAmount(DAI, 1)}
        trade={
          new InterfaceTrade({
            v2Routes: [],
            v3Routes: [testRoute],
            tradeType: TradeType.EXACT_INPUT,
          })
        }
      />
    )
    expect(component.getByText('Auto Router')).toBeTruthy()
    expect(component).toMatchSnapshot()
  })

  it('renders correctly, 2 hops', () => {
    const usdc = CurrencyAmount.fromRawAmount(USDC, 1)
    const dai = CurrencyAmount.fromRawAmount(DAI, 1)
    const component = renderComponent(
      <ToolbarOrderRouting
        gasUseEstimateUSD={CurrencyAmount.fromRawAmount(DAI, 1)}
        trade={
          new InterfaceTrade({
            v2Routes: [],
            v3Routes: [buildMultiV3Route(usdc, dai)],
            tradeType: TradeType.EXACT_INPUT,
          })
        }
      />
    )
    expect(component.getByText('Auto Router')).toBeTruthy()
    expect(component).toMatchSnapshot()
  })

  it('renders correctly, 3 hops across 2 swaps', () => {
    const usdc = CurrencyAmount.fromRawAmount(USDC, 1)
    const dai = CurrencyAmount.fromRawAmount(DAI, 1)
    const component = renderComponent(
      <ToolbarOrderRouting
        gasUseEstimateUSD={CurrencyAmount.fromRawAmount(DAI, 1)}
        trade={
          new InterfaceTrade({
            v2Routes: [],
            v3Routes: [buildMultiV3Route(usdc, dai), buildSingleV3Route(usdc, dai)],
            tradeType: TradeType.EXACT_INPUT,
          })
        }
      />
    )
    expect(component.getByText('Auto Router')).toBeTruthy()
    expect(component).toMatchSnapshot()
  })
})
