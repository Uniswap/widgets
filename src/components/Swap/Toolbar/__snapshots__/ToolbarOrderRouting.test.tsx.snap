// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarOrderRouting renders correctly, 1 hop 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div />
      <div
        style="isolation: isolate;"
      >
        <div
          class="Row-sc-1nzvhrh-0 ToolbarOrderRouting__OrderRoutingRow-sc-6rzp4w-0 iYjZBm cMNVqM"
        >
          <div
            class="type__TextWrapper-sc-16386l-0 hnjWoB body body-2 css-d5fsiy"
          >
            Order routing
          </div>
          <div
            class="Popover__Reference-sc-1liex6z-1 hplZxv"
          >
            <div
              class="Row-sc-1nzvhrh-0 hFfxta"
              style="outline: none;"
              tabindex="-1"
            >
              <svg
                class="icons-sc-lekdau-0 iSRwlR"
                fill="none"
                height="20"
                viewBox="0 0 23 20"
                width="23"
                xmlns="http://www.w3.org/2000/svg"
              >
                <defs>
                  <lineargradient
                    gradientTransform="rotate(95)"
                    id="AutoRouterIconGradient0"
                    x1="0"
                    x2="1"
                    y1="0"
                    y2="0"
                  >
                    <stop
                      id="stop1"
                      offset="0"
                      stop-color="#2274E2"
                    />
                    <stop
                      id="stop1"
                      offset="0.5"
                      stop-color="#2274E2"
                    />
                    <stop
                      id="stop2"
                      offset="1"
                      stop-color="#3FB672"
                    />
                  </lineargradient>
                </defs>
                <path
                  d="M16 16C10 16 9 10 5 10M16 16C16 17.6569 17.3431 19 19 19C20.6569 19 22 17.6569 22 16C22 14.3431 20.6569 13 19 13C17.3431 13 16 14.3431 16 16ZM5 10C9 10 10 4 16 4M5 10H1.5M16 4C16 5.65685 17.3431 7 19 7C20.6569 7 22 5.65685 22 4C22 2.34315 20.6569 1 19 1C17.3431 1 16 2.34315 16 4Z"
                  stroke="url(#AutoRouterIconGradient0)"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <div
                class="type__TextWrapper-sc-16386l-0 kEGyHj RoutingDiagram__StyledAutoRouterLabel-sc-3p4p27-0 hyAjPo css-aeawsu"
              >
                <div
                  class="type__TextWrapper-sc-16386l-0 jukanU subhead subhead-2 css-e17nie"
                >
                  Auto Router
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div />
    <div
      style="isolation: isolate;"
    >
      <div
        class="Row-sc-1nzvhrh-0 ToolbarOrderRouting__OrderRoutingRow-sc-6rzp4w-0 iYjZBm cMNVqM"
      >
        <div
          class="type__TextWrapper-sc-16386l-0 hnjWoB body body-2 css-d5fsiy"
        >
          Order routing
        </div>
        <div
          class="Popover__Reference-sc-1liex6z-1 hplZxv"
        >
          <div
            class="Row-sc-1nzvhrh-0 hFfxta"
            style="outline: none;"
            tabindex="-1"
          >
            <svg
              class="icons-sc-lekdau-0 iSRwlR"
              fill="none"
              height="20"
              viewBox="0 0 23 20"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  gradientTransform="rotate(95)"
                  id="AutoRouterIconGradient0"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="0"
                >
                  <stop
                    id="stop1"
                    offset="0"
                    stop-color="#2274E2"
                  />
                  <stop
                    id="stop1"
                    offset="0.5"
                    stop-color="#2274E2"
                  />
                  <stop
                    id="stop2"
                    offset="1"
                    stop-color="#3FB672"
                  />
                </lineargradient>
              </defs>
              <path
                d="M16 16C10 16 9 10 5 10M16 16C16 17.6569 17.3431 19 19 19C20.6569 19 22 17.6569 22 16C22 14.3431 20.6569 13 19 13C17.3431 13 16 14.3431 16 16ZM5 10C9 10 10 4 16 4M5 10H1.5M16 4C16 5.65685 17.3431 7 19 7C20.6569 7 22 5.65685 22 4C22 2.34315 20.6569 1 19 1C17.3431 1 16 2.34315 16 4Z"
                stroke="url(#AutoRouterIconGradient0)"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <div
              class="type__TextWrapper-sc-16386l-0 kEGyHj RoutingDiagram__StyledAutoRouterLabel-sc-3p4p27-0 hyAjPo css-aeawsu"
            >
              <div
                class="type__TextWrapper-sc-16386l-0 jukanU subhead subhead-2 css-e17nie"
              >
                Auto Router
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ToolbarOrderRouting renders correctly, 2 hops 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div />
      <div
        style="isolation: isolate;"
      >
        <div
          class="Row-sc-1nzvhrh-0 ToolbarOrderRouting__OrderRoutingRow-sc-6rzp4w-0 iYjZBm cMNVqM"
        >
          <div
            class="type__TextWrapper-sc-16386l-0 hnjWoB body body-2 css-d5fsiy"
          >
            Order routing
          </div>
          <div
            class="Popover__Reference-sc-1liex6z-1 hplZxv"
          >
            <div
              class="Row-sc-1nzvhrh-0 hFfxta"
              style="outline: none;"
              tabindex="-1"
            >
              <svg
                class="icons-sc-lekdau-0 iSRwlR"
                fill="none"
                height="20"
                viewBox="0 0 23 20"
                width="23"
                xmlns="http://www.w3.org/2000/svg"
              >
                <defs>
                  <lineargradient
                    gradientTransform="rotate(95)"
                    id="AutoRouterIconGradient2"
                    x1="0"
                    x2="1"
                    y1="0"
                    y2="0"
                  >
                    <stop
                      id="stop1"
                      offset="0"
                      stop-color="#2274E2"
                    />
                    <stop
                      id="stop1"
                      offset="0.5"
                      stop-color="#2274E2"
                    />
                    <stop
                      id="stop2"
                      offset="1"
                      stop-color="#3FB672"
                    />
                  </lineargradient>
                </defs>
                <path
                  d="M16 16C10 16 9 10 5 10M16 16C16 17.6569 17.3431 19 19 19C20.6569 19 22 17.6569 22 16C22 14.3431 20.6569 13 19 13C17.3431 13 16 14.3431 16 16ZM5 10C9 10 10 4 16 4M5 10H1.5M16 4C16 5.65685 17.3431 7 19 7C20.6569 7 22 5.65685 22 4C22 2.34315 20.6569 1 19 1C17.3431 1 16 2.34315 16 4Z"
                  stroke="url(#AutoRouterIconGradient2)"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <div
                class="type__TextWrapper-sc-16386l-0 kEGyHj RoutingDiagram__StyledAutoRouterLabel-sc-3p4p27-0 hyAjPo css-aeawsu"
              >
                <div
                  class="type__TextWrapper-sc-16386l-0 jukanU subhead subhead-2 css-e17nie"
                >
                  Auto Router
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div />
    <div
      style="isolation: isolate;"
    >
      <div
        class="Row-sc-1nzvhrh-0 ToolbarOrderRouting__OrderRoutingRow-sc-6rzp4w-0 iYjZBm cMNVqM"
      >
        <div
          class="type__TextWrapper-sc-16386l-0 hnjWoB body body-2 css-d5fsiy"
        >
          Order routing
        </div>
        <div
          class="Popover__Reference-sc-1liex6z-1 hplZxv"
        >
          <div
            class="Row-sc-1nzvhrh-0 hFfxta"
            style="outline: none;"
            tabindex="-1"
          >
            <svg
              class="icons-sc-lekdau-0 iSRwlR"
              fill="none"
              height="20"
              viewBox="0 0 23 20"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  gradientTransform="rotate(95)"
                  id="AutoRouterIconGradient2"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="0"
                >
                  <stop
                    id="stop1"
                    offset="0"
                    stop-color="#2274E2"
                  />
                  <stop
                    id="stop1"
                    offset="0.5"
                    stop-color="#2274E2"
                  />
                  <stop
                    id="stop2"
                    offset="1"
                    stop-color="#3FB672"
                  />
                </lineargradient>
              </defs>
              <path
                d="M16 16C10 16 9 10 5 10M16 16C16 17.6569 17.3431 19 19 19C20.6569 19 22 17.6569 22 16C22 14.3431 20.6569 13 19 13C17.3431 13 16 14.3431 16 16ZM5 10C9 10 10 4 16 4M5 10H1.5M16 4C16 5.65685 17.3431 7 19 7C20.6569 7 22 5.65685 22 4C22 2.34315 20.6569 1 19 1C17.3431 1 16 2.34315 16 4Z"
                stroke="url(#AutoRouterIconGradient2)"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <div
              class="type__TextWrapper-sc-16386l-0 kEGyHj RoutingDiagram__StyledAutoRouterLabel-sc-3p4p27-0 hyAjPo css-aeawsu"
            >
              <div
                class="type__TextWrapper-sc-16386l-0 jukanU subhead subhead-2 css-e17nie"
              >
                Auto Router
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ToolbarOrderRouting renders correctly, 3 hops across 2 swaps 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div />
      <div
        style="isolation: isolate;"
      >
        <div
          class="Row-sc-1nzvhrh-0 ToolbarOrderRouting__OrderRoutingRow-sc-6rzp4w-0 iYjZBm cMNVqM"
        >
          <div
            class="type__TextWrapper-sc-16386l-0 hnjWoB body body-2 css-d5fsiy"
          >
            Order routing
          </div>
          <div
            class="Popover__Reference-sc-1liex6z-1 hplZxv"
          >
            <div
              class="Row-sc-1nzvhrh-0 hFfxta"
              style="outline: none;"
              tabindex="-1"
            >
              <svg
                class="icons-sc-lekdau-0 iSRwlR"
                fill="none"
                height="20"
                viewBox="0 0 23 20"
                width="23"
                xmlns="http://www.w3.org/2000/svg"
              >
                <defs>
                  <lineargradient
                    gradientTransform="rotate(95)"
                    id="AutoRouterIconGradient4"
                    x1="0"
                    x2="1"
                    y1="0"
                    y2="0"
                  >
                    <stop
                      id="stop1"
                      offset="0"
                      stop-color="#2274E2"
                    />
                    <stop
                      id="stop1"
                      offset="0.5"
                      stop-color="#2274E2"
                    />
                    <stop
                      id="stop2"
                      offset="1"
                      stop-color="#3FB672"
                    />
                  </lineargradient>
                </defs>
                <path
                  d="M16 16C10 16 9 10 5 10M16 16C16 17.6569 17.3431 19 19 19C20.6569 19 22 17.6569 22 16C22 14.3431 20.6569 13 19 13C17.3431 13 16 14.3431 16 16ZM5 10C9 10 10 4 16 4M5 10H1.5M16 4C16 5.65685 17.3431 7 19 7C20.6569 7 22 5.65685 22 4C22 2.34315 20.6569 1 19 1C17.3431 1 16 2.34315 16 4Z"
                  stroke="url(#AutoRouterIconGradient4)"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <div
                class="type__TextWrapper-sc-16386l-0 kEGyHj RoutingDiagram__StyledAutoRouterLabel-sc-3p4p27-0 hyAjPo css-aeawsu"
              >
                <div
                  class="type__TextWrapper-sc-16386l-0 jukanU subhead subhead-2 css-e17nie"
                >
                  Auto Router
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div />
    <div
      style="isolation: isolate;"
    >
      <div
        class="Row-sc-1nzvhrh-0 ToolbarOrderRouting__OrderRoutingRow-sc-6rzp4w-0 iYjZBm cMNVqM"
      >
        <div
          class="type__TextWrapper-sc-16386l-0 hnjWoB body body-2 css-d5fsiy"
        >
          Order routing
        </div>
        <div
          class="Popover__Reference-sc-1liex6z-1 hplZxv"
        >
          <div
            class="Row-sc-1nzvhrh-0 hFfxta"
            style="outline: none;"
            tabindex="-1"
          >
            <svg
              class="icons-sc-lekdau-0 iSRwlR"
              fill="none"
              height="20"
              viewBox="0 0 23 20"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  gradientTransform="rotate(95)"
                  id="AutoRouterIconGradient4"
                  x1="0"
                  x2="1"
                  y1="0"
                  y2="0"
                >
                  <stop
                    id="stop1"
                    offset="0"
                    stop-color="#2274E2"
                  />
                  <stop
                    id="stop1"
                    offset="0.5"
                    stop-color="#2274E2"
                  />
                  <stop
                    id="stop2"
                    offset="1"
                    stop-color="#3FB672"
                  />
                </lineargradient>
              </defs>
              <path
                d="M16 16C10 16 9 10 5 10M16 16C16 17.6569 17.3431 19 19 19C20.6569 19 22 17.6569 22 16C22 14.3431 20.6569 13 19 13C17.3431 13 16 14.3431 16 16ZM5 10C9 10 10 4 16 4M5 10H1.5M16 4C16 5.65685 17.3431 7 19 7C20.6569 7 22 5.65685 22 4C22 2.34315 20.6569 1 19 1C17.3431 1 16 2.34315 16 4Z"
                stroke="url(#AutoRouterIconGradient4)"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <div
              class="type__TextWrapper-sc-16386l-0 kEGyHj RoutingDiagram__StyledAutoRouterLabel-sc-3p4p27-0 hyAjPo css-aeawsu"
            >
              <div
                class="type__TextWrapper-sc-16386l-0 jukanU subhead subhead-2 css-e17nie"
              >
                Auto Router
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
